src/dmd/cparse.d

```diff
@@ -2760,19 +2761,28 @@ final class CParser(AST) : Parser!AST
         {
             tag = token.ident;
             nextToken();
+            if (token.value == TOK.semicolon)
+            {
+                /* `struct tag;` always results in a declaration
+                 * in the current scope
+                 */
+                auto stag = (structOrUnion == TOK.struct_)
+                    ? new AST.StructDeclaration(loc, tag, false)
+                    : new AST.UnionDeclaration(loc, tag);
+                if (!symbols)
+                    symbols = new AST.Dsymbols();
+                symbols.push(stag);
+                return new AST.TypeStruct(stag);
```

Comment:
This pattern is also repeated at the end of the function, consider putting it into a nested function.

Reply:
done